pkg/api/meta/priority.go

```diff
@@ -0,0 +1,173 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package meta
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api/unversioned"
+)
+
+const (
+	AnyGroup    = "*"
+	AnyVersion  = "*"
+	AnyResource = "*"
+	AnyKind     = "*"
+)
+
+// PriorityRESTMapper is a wrapper for automatically choosing a particular Resource or Kind
+// when multiple matches are possible
+type PriorityRESTMapper struct {
+	// Delegate is the RESTMapper to use to locate all the Kind and Resource matches
+	Delegate RESTMapper
+
+	// ResourcePriority is a list of priority patterns to apply to matching resources.
+	// The list of all matching resources is narrowed based on the patterns until only one remains.
+	// A pattern with no matches is skipped.  A pattern with more than one match uses its
+	// matches as the list to continue matching against.
+	ResourcePriority []unversioned.GroupVersionResource
+
+	// KindPriority is a list of priority patterns to apply to matching kinds.
+	// The list of all matching kinds is narrowed based on the patterns until only one remains.
+	// A pattern with no matches is skipped.  A pattern with more than one match uses its
+	// matches as the list to continue matching against.
+	KindPriority []unversioned.GroupVersionKind
+}
+
+func (m PriorityRESTMapper) String() string {
+	return fmt.Sprintf("PriorityRESTMapper{\n\t%v\n\t%v\n\t%v\n}", m.ResourcePriority, m.KindPriority, m.Delegate)
+}
+
+// ResourceFor finds all kinds, then passes them through the ResourcePriority patterns to find a single matching hit.
+func (m PriorityRESTMapper) ResourceFor(resource unversioned.GroupVersionResource) (unversioned.GroupVersionResource, error) {
```

Comment:
Yes, absolutely. I'm not even sure why I mentioned it.

On Thu, Feb 18, 2016 at 10:55 AM, David Eads <notifications@github.com>
wrote:

> In pkg/api/meta/priority.go
> <https://github.com/kubernetes/kubernetes/pull/21272#discussion_r53362976>
> :
>
> > +	// matches as the list to continue matching against.
> > +	ResourcePriority []unversioned.GroupVersionResource
> > +
> > +	// KindPriority is a list of priority patterns to apply to matching kinds.
> > +	// The list of all matching kinds is narrowed based on the patterns until only one remains.
> > +	// A pattern with no matches is skipped.  A pattern with more than one match uses its
> > +	// matches as the list to continue matching against.
> > +	KindPriority []unversioned.GroupVersionKind
> > +}
> > +
> > +func (m PriorityRESTMapper) String() string {
> > +	return fmt.Sprintf("PriorityRESTMapper{\n\t%v\n\t%v\n\t%v\n}", m.ResourcePriority, m.KindPriority, m.Delegate)
> > +}
> > +
> > +// ResourceFor finds all kinds, then passes them through the ResourcePriority patterns to find a single matching hit.
> > +func (m PriorityRESTMapper) ResourceFor(resource unversioned.GroupVersionResource) (unversioned.GroupVersionResource, error) {
>
> If this is ever done in a hot loop, making a map from GVR to index # in the
> priority list would change this algorithm from O(N*M) to just O(N).
> Probably not worth even considering at this point.
>
> Not that I know of. The main user is the resource.Builder which usually
> gets called once in a kubectl client-side command. Because matches are
> winnowed out (doesn't have to be a single exclusive hit), you end up
> needing to know all matches (not just the first), so I think you still end
> up at O(N*slightly less than M).
>
> I'd also claim that the client side numbers of N and M are both small, so
> can we defer a performance enhancement?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21272/files#r53362976>.
>